plan/src/main/java/org/apache/flink/agents/plan/serializer/ActionJsonSerializer.java

```diff
@@ -0,0 +1,99 @@
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.agents.plan.serializer;
+
+import org.apache.flink.agents.api.Event;
+import org.apache.flink.agents.plan.Action;
+import org.apache.flink.agents.plan.JavaFunction;
+import org.apache.flink.agents.plan.PythonFunction;
+import org.apache.flink.shaded.jackson2.com.fasterxml.jackson.core.JsonGenerator;
+import org.apache.flink.shaded.jackson2.com.fasterxml.jackson.databind.SerializerProvider;
+import org.apache.flink.shaded.jackson2.com.fasterxml.jackson.databind.ser.std.StdSerializer;
+
+import java.io.IOException;
+
+/**
+ * Custom serializer for {@link Action} that handles the serialization of the function and event
+ * types.
+ */
+public class ActionJsonSerializer extends StdSerializer<Action> {
+    public ActionJsonSerializer() {
+        super(Action.class);
+    }
+
+    @Override
+    public void serialize(
+            Action action, JsonGenerator jsonGenerator, SerializerProvider serializerProvider)
+            throws IOException {
+        jsonGenerator.writeStartObject();
+
+        // Write name field
+        jsonGenerator.writeStringField("name", action.getName());
+
+        // Write exec field
+        if (action.getExec() instanceof JavaFunction) {
+            JavaFunction javaFunction = (JavaFunction) action.getExec();
+            serializeJavaFunction(jsonGenerator, javaFunction);
+        } else if (action.getExec() instanceof PythonFunction) {
+            PythonFunction pythonFunction = (PythonFunction) action.getExec();
+            serializePythonFunction(jsonGenerator, pythonFunction);
+        } else {
+            throw new IllegalArgumentException(
+                    "Unsupported function type: " + action.getExec().getClass().getName());
+        }
+
+        // Write listenEventTypes field
+        jsonGenerator.writeFieldName("listen_event_types");
+        jsonGenerator.writeStartArray();
+        for (Class<? extends Event> eventType : action.getListenEventTypes()) {
+            jsonGenerator.writeString(eventType.getName());
+        }
+        jsonGenerator.writeEndArray();
+
+        jsonGenerator.writeEndObject();
+    }
+
+    private void serializePythonFunction(JsonGenerator gen, PythonFunction func)
+            throws IOException {
+        gen.writeFieldName("exec");
+        gen.writeStartObject();
+        // Mark it as a Python function
+        gen.writeStringField("func_type", PythonFunction.class.getSimpleName());
```

Comment:
Define the `func_type` explicitly as static variable `PythonFunction` instead of relying on class name? Serde will break in case `PythonFunction` class gets refactored and change name